docs/sdk-configuration.md

```diff
@@ -0,0 +1,492 @@
+}
+interface SpanExporter extends OpenTelemetryComponent {
+}
+public class BatchExporter implements SpanExporter {
+  
+  private volatile Tracer tracer;
+  
+  @Override
+  public void setOpenTelemetry(OpenTelemetry openTelemetry) {
+    traer = openTelemetry.getTracerProvider().get("spanexporter");
+  }
+  
+  @Override
+  public void export() {
+    tracer.spanBuilder("export").startSpan();
+  }
+}
+public class OpenTelemetrySdkBuilder {
+  public OpenTelemetrySdkBuilder addSpanExporter(SpanExporter exporter) {
+    tracerProvider.addSpanExporter(exporter);
+    components.add(exporter);
+  }
+
+  public OpenTelemetrySdkBuilder setSampler(Sampler sampler) {
+    tracerProvider.setSampler(sampler);
+    components.add(sampler);
+  }
+
+  public OpenTelemetry build() {
+    OpenTelemetrySdk sdk = new OpenTelemetrySdk(tracerProvider.build(), meterProvider.build());
+    for (OpenTelemetryComponent component : components) {
+      component.setOpenTelemetry(sdk);
+    }
+  }
+}
+```
+
+A framework author will have an even easier time since most dependency injection frameworks
+natively support lazy injection.
+
+```java
+@Component
+public class MonitoringModule {
+  
+  @Bean
+  @ForSpanExporter
+  public Tracer tracer(TracerProvider tracerProvider) {
+    return tracerProvider.get("spanexporter");
+  }
+}
+
+@Component
+public class MyExporter implements SpanExporter {
+
+  private Lazy<Tracer> tracer;
+  
+  @Inject
+  public MyExporter(@ForSpanExporter Lazy<Tracer> tracer) {
+    this.tracer = tracer;
+  }
+  
+  @Override
+  public void export() {
+    tracer.get().spanBuilder("export").startSpan();
+  }
+}
+```
+
+## Immutability of OpenTelemetry configuration
+
+The above attempts to avoid allowing a built SDK to be mutated. Allowing mutation can make code
+harder to reason about (any component, even deep in business logic, could update the SDK without
+hindrance), can reduce performance (require volatile read on most operations), and produce thread
+safety issues if not well implemented. In particular, compared to the current state as of writing,
+
+- `addSpanProcessor` is not needed because we instead push the complexity of handling telemetry
+within telemetry components to those components, where the maintainers will have more domain
+knowledge. It allows this mutator method to be removed from the end-user API.
+
+- `updateTraceConfig` - instead of allowing updates at the top level, we should consider making
+`TraceConfig` an interface, and the SDK default implementation is static. It allows the above
+benefits of immutability to be in place for the common case where dynamic updates are not needed.
+Where dynamic updates are needed, it can be replaced with a mutable implementation instead of making
+the SDK configuration mutable. This keeps update methods out of the end-user APIs and will generally
+give framework developers more control by handling dynamicism themselves without the chance of
+end-users to affect it negatively.
```

Comment:
If TraceConfig is not "immutable" you have hard time and undefined behaviors. What does it happen if we decide to call sampler twice and gets changed between the calls? Or any other property that we use?

Reply:
By making immutable we can guarantee that one config is applied for the entire lifetime of one span, and you cannot do this with a mutable config.

You may say that you add a "Factory" that returns an immutable TraceConfig, that is a lot of unnecessary allocations (you may cache the result). But if there is a factory why someone cannot do the same thing and update the Factory?

Reply:
How are we going to support dynamic Samplers?

Reply:
Sampler is already an interface I was always expecting them to be a dynamic implementation of it.

Reply:
All current Samplers are immutable and are changed by swapping instances.

Reply:
sure, but that doesn't mean that other Samplers could be more dynamic (or we could even update our existing Samplers to be dynamic without altering the interface, I think).